lib/rules/function-component-definition.js

```diff
@@ -21,8 +21,8 @@ function buildFunction(template, parts) {
 
 const NAMED_FUNCTION_TEMPLATES = {
   'function-declaration': 'function {name}{typeParams}({params}){returnType} {body}',
-  'arrow-function': 'var {name}{typeAnnotation} = {typeParams}({params}){returnType} => {body}',
-  'function-expression': 'var {name}{typeAnnotation} = function{typeParams}({params}){returnType} {body}',
+  'arrow-function': 'const {name}{typeAnnotation} = {typeParams}({params}){returnType} => {body}',
+  'function-expression': 'const {name}{typeAnnotation} = function{typeParams}({params}){returnType} {body}',
```

Comment:
This is intentional; the prefer-const and no-var core rules handle changing these to const.

Reply:
> This is intentional; the prefer-const and no-var core rules handle changing these to const.

Unfortunately not for auto-fixing, those rules just signal it as an error

![eslint_react](https://user-images.githubusercontent.com/31244713/159171112-8158574f-7004-43b7-a46a-4c12a2b1a9e7.gif)

Reply:
@ljharb I've noticed that the autofix for this rule also will change `const SomeComponent = () =>` into `var SomeComponent = () =>` even though this should be already valid and not fixed.

As I was thinking a bit more last night that this does depend on the developers preference - some prefer `var` some prefer `const` and some prefer `let`. Would you be adverse to the following that is more verbose than current, and does have a bit of overlap with the `prefer-const` and `no-var` type rules. But it would allow us to set up this rule so that it doesn't conflict with other rules.

* `arrow-function`, `function-expression`  - dynamically figure if we should use let/const/var according to the logic outlined in this pull

* `arrow-function-let`, `function-expression-let` - force let

* `arrow-function-const`, `function-expression-const` - force const

* `arrow-function-var`, `function-expression-var` - force var

For `no-var` and `prefer-const` here's a eslint playground demonstrating that it can't fix this case https://eslint.org/demo#eyJ0ZXh0IjoiZXhwb3J0IHZhciBmb28gPSAoKSA9PiA8ZGl2PkhlbGxvIFdvcmxkPC9kaXY+OyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7InByZWZlci1jb25zdCI6Miwibm8tdmFyIjoyfSwiZW52Ijp7fX19

Reply:
Ok, let’s see.

1. I think it’s 100% a bug if it’s changing an existing const/let in the proper form to a var in the proper form; that should be fixed regardless (ideally in a separate PR).
1. I think the dynamic logic with the current options is both a strict improvement and also would likely cover 99.999999% of use cases.
1. thus, I’d rather evaluate the remaining use cases after the first two are done :-)

Reply:
Sure, I have fixed the bug as part of the code I wrote to do the const/var detection and added tests for it too. (See comment in main thread) 

I think for me personally it's not going to be without the extra options so I have no strong need for those extra options at present.

Reply:
If I didn't do any mistake, your branch should be merged into mine @SimeonC